### PR TITLE
[SKIA] UseOpacitySaveLayer feature switch

### LINKED ISSUE #9227: Opacity doesn't work like in WPF
**Describe the bug**

I'm coming from a WPF background.  I was trying to use a semi-transparent complex element by setting `Opacity` to < `1`.  While this worked, the result is not expected since setting `Opacity` on the root of a complex element seems to make each descendent element get the opacity applied instead of rendering the final result of the complex element at a certain opacity.  

See screens below for a very basic example, but keep in mind that in my case, I have a much more complex element that looks awful when all of its parts are showing through each other.

**To Reproduce**

My test code in both WPF and Avalonia is this where the top half is a fully opaque set of ellipses and the bottom half is the same, but has the containing `StackPanel` set to 50% opacity:

```
<Grid Height="200" HorizontalAlignment="Center" VerticalAlignment="Center" Background="Silver">
	<StackPanel Margin="10" Orientation="Horizontal" 
		HorizontalAlignment="Center" VerticalAlignment="Top">
		<Ellipse Margin="0,0,-20,0" Width="80" Height="80" Fill="Green" />
		<Ellipse Margin="-20,0,0,0" Width="80" Height="80" Fill="Blue" />
	</StackPanel>
	<StackPanel Margin="10" Orientation="Horizontal" 
		HorizontalAlignment="Center" VerticalAlignment="Bottom" Opacity="0.5">
		<Ellipse Margin="0,0,-20,0" Width="80" Height="80" Fill="Green" />
		<Ellipse Margin="-20,0,0,0" Width="80" Height="80" Fill="Blue" />
	</StackPanel>
</Grid>
```

The unexpected result in Avalonia is:
image

You can clearly see how the individual `Ellipse` shapes are all partially transparent, which is very bad when you have a complex element and set opacity on it.

**Expected behavior**

The expected result as in WPF is:
image

You can see here the partially transparent `StackPanel` has its final rendering in tact, but just faded into the background.  This is what I would expect in Avalonia too.

Is there any way to achieve this?  Otherwise, `Opacity` is not practical for use.

Even if there is a workaround, I would have expected the WPF-like behavior out of the box.

**Desktop (please complete the following information):**

- OS: Windows
- Version 11.0.0-preview2

### PR DESCRIPTION
**Pull request by Gillibald**
## What does the pull request do?

This PR introduces a feature switch to enable opacity handling via SKCanvas.SaveLayer

To enable this feature you need to add `.With(new SkiaOptions { UseOpacitySaveLayer = true });` to your AppBuilder setup

```
  <Grid Height="200" HorizontalAlignment="Center" VerticalAlignment="Center" Background="Silver">
	<StackPanel Margin="10" Orientation="Horizontal" 
		HorizontalAlignment="Center" VerticalAlignment="Top">
		<Ellipse Margin="0,0,-20,0" Width="80" Height="80" Fill="Green" />
		<Ellipse Margin="-20,0,0,0" Width="80" Height="80" Fill="Blue" />
	</StackPanel>
	<StackPanel Margin="10" Orientation="Horizontal" 
		HorizontalAlignment="Center" VerticalAlignment="Bottom" Opacity="0.5">
		<Ellipse Margin="0,0,-20,0" Width="80" Height="80" Fill="Green" />
		<Ellipse Margin="-20,0,0,0" Width="80" Height="80" Fill="Blue" />
	</StackPanel>
</Grid>
```
### UseOpacitySaveLayer=true
Screenshot 2023-02-09 104045
### UseOpacitySaveLayer=false
Screenshot 2023-02-09 104112

## Checklist

- [ ] Added unit tests (if possible)?
- [ ] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Documentation with user documentation

## Fixed issues

Fixes: #9227

### CODE REVIEW

File: tests/Avalonia.RenderTests/Controls/TextBlockTests.cs [hunk 1]
```diff
@@ -92,7 +92,7 @@ Control text(VerticalAlignment verticalAlignment, bool clip = true, bool restric
             };
 
             await RenderToFile(target);
-            CompareImages();
+            CompareImages(skipImmediate: true);
```
Comment (kekekeks) [inline tests/Avalonia.RenderTests/Controls/TextBlockTests.cs:95]: Does this PR somehow breaks immediate mode rendering? It might be worth investigating.

Comment (Gillibald) [reply tests/Avalonia.RenderTests/Controls/TextBlockTests.cs:95]: It seems that software rendering produces slightly different font rendering not sure why. I will try different edging modes. Ideally  immediate rendering would not use BitmapRenderTarget

**Comment by Mikolaytis** [comment]
PushLayer is an operation that easts a lot of performance.
AFAIK Every call creates a texture to hold all draw data. Pop applies texture data to initial canvas and drops away a texture.
1. Before accepting this - test is well. I think it's not worth it.
2. If you still want to use it. Use it only when a layer is complicated. Using it for rectangles is a bad idea. 
3. Also if you will use it - I think you want to create as small buffer textures as possible - so maybe you need to provide bounds property to SaveLayer.

File: tests/Avalonia.RenderTests/Controls/TextBlockTests.cs (hunk 1 above)
Comment (kekekeks) [reply tests/Avalonia.RenderTests/Controls/TextBlockTests.cs:95]: Render tests are always using software rendering

**Comment by Gillibald** [comment]
The only option I see to avoid an oversized back buffer is to use DrawingContext.CreateLayer for opacity handling

**Comment by Gillibald** [comment]
PushOpacityMask is implemented in the same way. So there is at least one additional code path that is suffering from this performance hit.

**Comment by kekekeks** [comment]
Due to performance considerations using SaveLayer is a bit more involved than just changing the Skia context implementation, we also need changes on the renderer side so it would only create a layer when a node with opacity has overlapping children

e. g.
```xml
<Border Opacity="0.3">
   <StackPanel>
      <Border Background="Red" Width="10" Height="10" />
   </StackPanel>
</Border>
```

should _not_ call SaveLayer, since we can just adjust the rectangle color like we already do.

```xml
<Border Opacity="0.3">
   <StackPanel>
      <Border Background="Red" Width="10" Height="10" Padding="10" />
      <Border Background="Blue" Width="10" Height="10" Padding="10" />
   </StackPanel>
</Border>
```

doesn't need a layer either since there is no overlap.


This example should only have one SaveLayer call:
```xml
<Border Opacity="0.3">
   <StackPanel Opacity="0.5">
      <Button>Test</Button>
   </StackPanel>
</Border>
```
`StackPanel` here has overlapping children, but by itself it's a primitive drawing node since it would use `SaveLayer` for its contents already. So no `SaveLayer` call is needed for `Border`


So we need both "virtual" and "layered" opacity support in the drawing context with renderer deciding which one to use for a particular node

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0029793-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0029840-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

File: src/Skia/Avalonia.Skia/ISkiaSharpApiLeaseFeature.cs [hunk 2]
```diff
@@ -16,5 +16,4 @@ public interface ISkiaSharpApiLease : IDisposable
     SKCanvas SkCanvas { get; }
     GRContext GrContext { get; }
     SKSurface SkSurface { get; }
-    double CurrentOpacity { get; }
```
Comment (maxkatz6) [inline src/Skia/Avalonia.Skia/ISkiaSharpApiLeaseFeature.cs:19]: As per discussion, this property shouldn't be removed from the API.

**Review by maxkatz6 — CHANGES_REQUESTED** [review]

**Review by kekekeks — APPROVED** [review]

**Review by kekekeks — CHANGES_REQUESTED** [review]
`CurrentOpacity` is still needed. For layer based opacity we can just always return 1.

**Comment by avaloniaui-team** [comment]
You can test this PR using the following package version. `11.0.999-cibuild0030285-beta`. (feed url: https://pkgs.dev.azure.com/AvaloniaUI/AvaloniaUI/_packaging/avalonia-all/nuget/v3/index.json) [PRBUILDID]

File: src/Avalonia.Base/Rendering/Composition/Server/ServerCompositionVisual.cs [hunk 3]
```diff
@@ -54,15 +54,16 @@ public void Render(CompositorDrawingContextProxy canvas, Rect currentTransformed
             var transform = GlobalTransformMatrix;
             canvas.PostTransform = MatrixUtils.ToMatrix(transform);
             canvas.Transform = Matrix.Identity;
-            
-            var boundsRect = new Rect(new Size(Size.X, Size.Y));
+
+            if (Opacity != 1)
+                canvas.PushOpacity(Opacity, boundsRect);
```
[the remark this answers is not in the document]
Comment (Gillibald) [reply src/Avalonia.Base/Rendering/Composition/Server/ServerCompositionVisual.cs:59]: boundsRect is only used to ask Skia for a smaller background surface which isn't guaranteed to be equal in size. The rendering result isn't clipped on `Restore`. I guess I have to write a unit test to make sure nothing is clipped.

**Review by kekekeks — APPROVED** [review]

**Comment by gerhard17** [comment]
Shouldn't `ImmediateDrawingContext.PushOpacity(double opacity, Rect bounds)` have a nullable Rect? argument?

**Pull request merged**